gt/main.c

```diff
@@ -2017,3 +2036,74 @@ l_update_gt_lua_states(lua_State *l)
 
 	return 0;
 }
+
+int
+l_update_gt_lua_states_incrementally(lua_State *l)
+{
+	int i;
+	uint32_t ctypeid;
+	struct gt_config *gt_conf;
+	uint32_t correct_ctypeid_gt_config = luaL_get_ctypeid(l,
+		CTYPE_STRUCT_GT_CONFIG_PTR);
+	size_t len;
+	const char *lua_bytecode;
+
+	/* First argument must be of type CTYPE_STRUCT_GT_CONFIG_PTR. */
+	void *cdata = luaL_checkcdata(l, 1,
+		&ctypeid, CTYPE_STRUCT_GT_CONFIG_PTR);
+	if (ctypeid != correct_ctypeid_gt_config)
+		luaL_error(l, "Expected `%s' as first argument",
+			CTYPE_STRUCT_GT_CONFIG_PTR);
+
+	gt_conf = *(struct gt_config **)cdata;
+
+	/* Second argument must be a Lua bytecode. */
+	lua_bytecode = lua_tolstring(l, -1, &len);
+	if (lua_bytecode == NULL || len == 0)
+		luaL_error(l, "gt: invalid lua bytecode\n");
+
+	if (lua_gettop(l) != 2)
+		luaL_error(l, "Expected two arguments, however it got %d arguments",
+			lua_gettop(l));
+
+	for (i = 0; i < gt_conf->num_lcores; i++) {
+		int ret;
+		struct gt_instance *instance = &gt_conf->instances[i];
+		unsigned int lcore_id = gt_conf->lcores[i];
+		struct gt_cmd_entry *entry;
+		void *lua_bytecode_buff = rte_malloc_socket("lua_bytecode",
+			len, 0, rte_lcore_to_socket_id(lcore_id));
+		if (lua_bytecode_buff == NULL) {
+			luaL_error(l, "gt: failed to send new lua update chunk bytecode to GT block %u at lcore %u due to failure of allocating memory\n",
+				i, lcore_id);
+
+			continue;
+		}
+
+		rte_memcpy(lua_bytecode_buff, lua_bytecode, len);
```

Comment:
Move `rte_memcpy()` immediately after line `entry->u.bc.lua_bytecode = lua_bytecode_buff;` below, so we only go through the trouble of copying memory if we have all the necessary memory.